### PR TITLE
Feature 64-byte FDCAN messages

### PR DESCRIPTION
**Pull request by rasmuskleist**
The code has been tested on a custom board using the stm32g491met6 chip. There are a few points that should be considered.

1. The LongMessage is also currently available to devices only supporting CAN. Naturally guards can be implemented in lbuild to only generate LongMessage if supported. However, only Fdcan{{ id }} implements methods for sending LongMessage.
2. LongMessage has a 64-byte buffer but there is not directly implemented support for other buffer sizes like 32-byte. The MessageBase class is implemented such that the user can specify the buffer size accordingly.
3. The message queue now per default contains LongMessages when using FDCAN. Thereby the buffer is approximately eight times bigger as compared to CAN. Possibly it would make sense to change the default fdcan buffer size to accommodate this.
4. The FDCAN test is effectively the same as the can test for the nucleo g474. Hence both regular CAN and FDCAN could be tested in the original test file. 

I would like to hear your thoughts on these problems and if there would be other considerations?

### CODE REVIEW

File: examples/nucleo_g474re/fdcan/project.xml
Comment (rleh) [inline examples/nucleo_g474re/fdcan/project.xml]: `modm:platform:can:3` is not used.

File: examples/nucleo_g474re/fdcan/project.xml
Comment (rleh) [inline examples/nucleo_g474re/fdcan/project.xml]: `modm:platform:uart:2` is not used.

File: src/modm/architecture/interface/can_message.hpp
Comment (rleh) [inline src/modm/architecture/interface/can_message.hpp]:
```suggestion
/// @ingroup modm_architecture_can
template <uint8_t capacity>
struct MessageBase
```

File: src/modm/architecture/interface/can_message.hpp
Comment (rleh) [inline src/modm/architecture/interface/can_message.hpp]: Indentation: Here (and in some other places) you used spaces instead of tabs.
modm always uses tabs, please fix the indentation. (Sorry...)

File: src/modm/platform/can/stm32-fdcan/can.cpp.in
Comment (rleh) [inline src/modm/platform/can/stm32-fdcan/can.cpp.in]: Unused; and: This include might not be always available.

File: src/modm/architecture/interface/can_message.hpp
Comment (rleh) [inline src/modm/architecture/interface/can_message.hpp]: `setLength(64)` would look fine, but fail silently.

Maybe `setLength()` should take a length as an argument and convert is to the (next bigger) DLC?
And we could add a `setDlc()` function for the user to explicitly set the DLC?

File: src/modm/platform/can/stm32-fdcan/message_ram.hpp
Comment (rleh) [inline src/modm/platform/can/stm32-fdcan/message_ram.hpp]: Should be removed, see above.

**Review by rleh — COMMENTED** [review]
Nice work, thank you a lot!! :)

**Comment by rleh** [comment]
> 1. The LongMessage is also currently available to devices only supporting CAN. Naturally guards can be implemented in lbuild to only generate LongMessage if supported.

It does not really hurt performance or memory usage if it is not used, so I'm fine with having `LongMessage` always available.

> 2. LongMessage has a 64-byte buffer but there is not directly implemented support for other buffer sizes like 32-byte. The MessageBase class is implemented such that the user can specify the buffer size accordingly.

Nice 👍🏽

> 3. The message queue now per default contains LongMessages when using FDCAN. Thereby the buffer is approximately eight times bigger as compared to CAN. Possibly it would make sense to change the default fdcan buffer size to accommodate this.

I don't like this so much because I know some applications where the FDCAN IP is only used with traditional 8 byte messages and the larger buffers will probably cause problems.

Maybe we can have an lbuild option (per FDCAN instance) to enable the long messages? If the option defaults to `false` this PR should also be perfectly backward compatible.

**Comment by rasmuskleist** [comment]
Thanks for your comments!

I will make the code backward compatible. How do you suggest we do this? What seems to cause the problem is the depreciation of the length variable. For CAN the dlc and length is the same and hence for backward compatibility we could rename the dlc variable to length? I personally would prefer not to have both a dlc and a length variable. Also I much prefer using the dlc because it reduces the conversion to looking up in an array!

Also I like the idea of passing the message buffer length to lbuild. Then I suggest we remove the message base class template and set the message buffer size with lbuild so we don't have double templates. For example we add an option
`<option name="modm:can:message.buffer_size">64</option> `
or equivalent. Thereby we will only have to change the message code and there will be no need differentiate between Message and LongMessage in the CAN code. With this approach we should decide if we want to guard if the input given in the project file makes sense. For example so it's contained within the list of possible message lengths. However, passing a number would at least not break the code.

What do you think?

File: src/modm/architecture/interface/can_message.hpp
Comment (rasmuskleist) [reply src/modm/architecture/interface/can_message.hpp]: Sorry I will fix that

File: src/modm/architecture/interface/can_message.hpp
Comment (rasmuskleist) [reply src/modm/architecture/interface/can_message.hpp]: I was considering doing this, but I am also very reluctant to introduce more logic than necessary. However, since the code must be backward compatible this might be a better solution.

File: src/modm/platform/can/stm32-fdcan/message_ram.hpp
Comment (rasmuskleist) [reply src/modm/platform/can/stm32-fdcan/message_ram.hpp]: Yup, I will also remove that

**Comment by rasmuskleist** [comment]
I have made the change so the message buffer size can be set in project file. However, the change is global across all fdcan lines. I think it would be a nice addition to be able to specify it for the particular line. I can implement this by reintroducing the template parameter and allow the specific fd can line to send every message with buffer length less than or equal to the specified message buffer size. What do you think about that?

One should be very aware that the message used by the specific can line would then end up in the specific can line namespace with this approach e.g. Fdcan1::message.

**Comment by rasmuskleist** [comment]
I consider CAN FD done and use it for my own project already. What are your thoughts?

**Comment by rleh** [comment]
Sorry for the long delay, I was busy working on other projects.
I have already checked out your branch locally on will review it later or tomorrow.

**Comment by rasmuskleist** [comment]
Thank you! Let me know if you think further changes are necessary?

**Review by rleh — APPROVED** [review]
Nice work!

**Comment by rleh** [comment]
@rasmuskleist Do you want to squash and rebase your commits?

File: src/modm/architecture/module.lb
Comment (chris-durand) [inline src/modm/architecture/module.lb]:
```suggestion
                minimum=8, maximum=64,
```
Reducing the buffer size below 8 will cause undefined behaviour with many CAN drivers (e.g [here](https://github.com/modm-io/modm/blob/d125e6a31394fdb5c49ce2f1faabcdc162867e34/src/modm/platform/can/stm32/can.cpp.in#L214)). Having 8 as the minimum configurable capacity will be fine for now.

File: src/modm/architecture/interface/can_message.hpp.in
Comment (chris-durand) [inline src/modm/architecture/interface/can_message.hpp.in]: This long message handling is a pessimization for regular CAN with the default capacity of 8. The compiler does not understand it can optimize away the loop iterations and branches. We could do some compile-time handling with `if constexpr` to optimize for the common case:
```
if constexpr (capacity <= 8) {
    length = dlc = std::min(inLength, capacity);
} else {
    …;
}
```

File: src/modm/architecture/interface/can_message.hpp.in
Comment (chris-durand) [inline src/modm/architecture/interface/can_message.hpp.in:167]: Conventional C++ wisdom says these shouldn't be public. The class has to maintain an invariant on `dlc` and `length`. If one is changed, the other one has to be updated and not all values are valid. That is exactly what the `setLength()` and `setDataLengthCode()` member functions do.

Changing them to be private would be a breaking change, though. I am not really sure what the best option is. We have to make a choice between compatibility and sanity. What do you think @salkinium @rleh ?

File: src/modm/architecture/interface/can_message.hpp.in
Comment (chris-durand) [inline src/modm/architecture/interface/can_message.hpp.in]: Before the change specifying `inLength > 8` truncated the message length to 8. Now this will result in out-of-bounds memory accesses. Is this intended to work somehow?

Just as a side-note: Messages created using this constructor have a 11-bit standard id by default, the other constructor uses 29 bit extended. I would not change it because that could break perfectly fine user code at runtime.

File: src/modm/architecture/interface/can_message.hpp.in
Comment (chris-durand) [inline src/modm/architecture/interface/can_message.hpp.in]: Is this meant to handle sizes greater than 8? This change seems a bit odd. `dlcConversionTable[length]` is equal to `length` for all values that are valid length values. `length > 8` would always cause out-of-bounds memory accesses reading from a `uint64_t`.

**Review by chris-durand — COMMENTED** [review]
Thanks!

File: src/modm/architecture/interface/can_message.hpp.in
Comment (rasmuskleist) [reply src/modm/architecture/interface/can_message.hpp.in:167]: Yes, conventional C++ wisdom says these shouldn't be public. I have chosen to make both public because length was previously public. I personally, would prefer to make both private although this would interfere with backward compatibility.

Comment (rasmuskleist) [reply src/modm/architecture/interface/can_message.hpp.in:167]: Here @rleh mentions bacward compatability as an issue here:
[https://github.com/modm-io/modm/pull/882#issuecomment-1181084541](url)

File: src/modm/architecture/interface/can_message.hpp.in
Comment (rasmuskleist) [reply src/modm/architecture/interface/can_message.hpp.in]: I have changed the code to the previous state, where the in length is truncated at 8 and added a new constructor

File: src/modm/architecture/interface/can_message.hpp.in
Comment (rasmuskleist) [reply src/modm/architecture/interface/can_message.hpp.in]: I have resolved this by reverting the change to the previous version

File: src/modm/architecture/interface/can_message.hpp.in
Comment (chris-durand) [reply src/modm/architecture/interface/can_message.hpp.in:167]: We will not change this in this PR and leave the values as `public` for now.

**Comment by chris-durand** [comment]
@rasmuskleist Thank you very much for all the work and sorry review took such a long time.
We intend to merge the changes very soon. Since I don't have permission to push to your branch I'll open a new pull request soon with the commits cleaned up a bit and we'll merge shortly.

**Comment by rasmuskleist** [comment]
@chris-durand sorry for stagnating the development of this PR! I have been exceptionally busy with my thesis. I am currently facing some issues when the header type of the message is not extended. When extended everything works fine, however, when not extended it seems that the message is being sent but not received. Until now I have not resolved the issue but plan to do so in the near future. Have you got any ideas as to why this might be the case?

**Comment by rleh** [comment]
Sounds to me like the CAN message filter is not configured correctly.

**Comment by rasmuskleist** [comment]
Turned out it was the filter... from my part this pr is ready for merging now

**Comment by chris-durand** [comment]
@rasmuskleist @twast92 There is still a small issue left before we can merge. The CAN message header fails to compile on Mac OS:

```
modm/src/modm/architecture/interface/can_message.hpp:193:37: error: no matching function for call to 'begin(const uint8_t [8])'
  193 |                 std::copy(std::begin(rhs.data), std::end(rhs.data), std::begin(this->data));
```

The matching overload of `std::begin` is not found. This looks like an include problem. On other compiler versions / platforms the right file gets included indirectly somewhere. Could you try adding `#include <array>` in the header file and see if it resolves the CI error? Otherwise we should be ready to merge.

**Comment by chris-durand** [comment]
@twast92 That worked, thanks. Please squash the fix commit and we'll merge.

**Review by chris-durand — APPROVED** [review]

**Review by salkinium — APPROVED** [review]
Thanks!

**Pull request merged**
